ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+worker_2.register_activity(MyActivity, name: 'my-activity') # uses 'my-activity' as a name
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
+- In case connection sharing is a concern we can create and memoize a connection on the instance of
+  Temporal::Configuration
+
+And here is the `Worker` interface:
+
+```ruby
+class Worker
+  # Create a new instance of a Worker
+  def self.new(config: Temporal::Configuration, namespace: String, task_queue: String, **options) -> Worker
+
+  # Register a workflow with an optional :name
+  def register_workflow(Class, name: nil) -> void
+
+  # Register an activity with an optional :name
+  def register_activity(Class, name: nil) -> void
+
+  # Start the worker and block until the process exits
+  def run() -> void
+
+  # Start the worker asynchronously
+  def start() -> void
+
+  # Gracefully stop the worker (wait for all running workflow/activities to finish)
+  def stop() -> void
+end
+```
+
+Internally each worker will create a thread for task polling as well as a thread pool of a given
+size (passed to the initializer as an optional `:threads` argument). Then each activity task will be
+dispatched to an available thread (via an in-memory queue). To make sure a polling worker always has
+processing capacity a polling request will only be dispatched if there is an available thread in the
+pool (in other words the pool will have no buffer).
+
+A very similar model will be used for workflow tasks, with an added complexity of wrapping each
+workflow in its own Fiber (inside a thread). This serves two purposes:
+
+1. A workflow code can be paused and resumed during the history replay
+2. Sticky workflows can be left paused while waiting for the next workflow task
+
+Given that Fibers are very lightweight this will allow us to keep many stick workflows efficiently
+in memory and manage accordingly to the worker configuration.
+
+Here's a diagram to demonstrate the whole setup:
+
+<img src="./ruby_threads.png" width="400" align="center" alt="Temporal" />
+
+### Alternatives considered
+
+We've decided against using the `Client` to initialise the `Worker` (something that Go and Python
+SDKs are doing). It seemed like a wrong abstraction and would require making many attributes public
+on the client to make sure the worker can access them (expanding the public API of a client). Using
+a configuration object solves this and opens up further optimisation opportunities (that will be
+proposed in the 2nd phase).
```

Comment:
By deciding against this, you are gonna require a new server connection for every worker? Even if I make many dozens? Some people create short-lived workers ephemerally. You're gonna need every bit of client configuration for a worker anyways then, so you aren't saving anything, you're just making me configure the same thing twice.

Simply put: Everything that needs to communicate with Temporal servers needs a client. Whether you hide that from users making this unnecessary separation doesn't change the fact that there is still a client involved somewhere. Ideally for dev friendliness we encourage reuse.

Reply:
I think we might have a definition misunderstanding 😃 Here's how I see it:

- Connection — a gRPC client that provides low level protobuf based communication with Temporal
- Client — a high-level interface for SDK users that exposes certain methods for interaction with Temporal server via a connection
- Worker — activity/workflow processor that uses low-level connection for polling tasks and responding to them

So while there is connection sharing between a Client and a Worker, I don't agree that a Worker requires a Client unless my terminology above is incorrect.

Following your logic I'd expect Client to expose `poll_for_workflow_task`, `respond_workflow_task_completed`, etc methods for the Worker to use. Otherwise you're just using Client as a sharing mechanism for a connection, which doesn't feel right to me.

What am I missing? 🤔

Reply:
Yes, our terms may have been mixed up. Working w/ your terms...

> So while there is connection sharing between a Client and a Worker, I don't agree that a Worker requires a Client unless my terminology above is incorrect.

Can you clarify this connection sharing and how it's done and where the connections are stored? If a client can't own its connection, something has to. And the more you hide that from the user, the more confusing it is.

Also, everything that you need to build a "connection" for a "client" you also need to build a "connection" for the "worker", and since "client" is just thin stateless sugar over a "connection", why wouldn't you let users just reuse that known object to create the worker? Why hide it all? What benefit are you getting from not allowing the user to explicitly provide the "client" to the "worker" but instead have some hidden "connection" concept?

Reply:
> Can you clarify this connection sharing and how it's done and where the connections are stored?

Both a worker and a client retain a pointer to a connection from the configuration object once initialized.

The benefit is better composability — as you said a client is a thin layer of sugar over the connection. The worker doesn't care about it, it just needs a connection. Using a client to pass connection over is awkward in cases where you don't need a client, but only need a worker (which is probably most of the cases when you need a worker).

As discussed sync — this is pending me learning more about the Core and potentially revisiting some of these decisions.

Reply:
> The worker doesn't care about it, it just needs a connection. Using a client to pass connection over is awkward in cases where you don't need a client, but only need a worker (which is probably most of the cases when you need a worker).

To clarify, the worker needs a gRPC client. Specifically a core client. Since the Ruby client is just the Ruby class representing the gRPC/core client, it's not awkward that said representation is passed around where needed.

Reply:
Duplicate discussion. Use https://github.com/temporalio/proposals/pull/57#discussion_r866184882